### ISSUE TITLE
Linking error in qlite3/3.39.4 because it is without libm

### ISSUE THREAD

**Issue by yushb0602** [conan-io/conan#12273]
In  sqlite3/3.39.4 the lastest version, please add libm to shell.c. in  /home/ogic/.conan/data/sqlite3/3.39.4/_/_/source/CMakeLists.txt line 170

 target_link_libraries(sqlite3-bin PRIVATE ${PROJECT_NAME} **m**)

otherewise there are  linking errors on linux

sqlite3/3.37.2: CMake command: cmake --build "/home/ogic/.conan/data/sqlite3/3.37.2/_/_/build/e32c1f48bfda04d1f3ad0456e7afb1864c5ecabb/build/Release" '--' '-j2'
[ 25%] Building C object CMakeFiles/sqlite3.dir/src/sqlite3.c.o
[ 50%] Linking C static library libsqlite3.a
[ 50%] Built target sqlite3
[ 75%] Building C object CMakeFiles/sqlite3-bin.dir/src/shell.c.o
[100%] Linking C executable sqlite3
/usr/lib/gcc-cross/aarch64-linux-gnu/11/../../../../aarch64-linux-gnu/bin/ld: libsqlite3.a(sqlite3.c.o): in function `logFunc':
sqlite3.c:(.text+0x356c8): undefined reference to `log'
/usr/lib/gcc-cross/aarch64-linux-gnu/11/../../../../aarch64-linux-gnu/bin/ld: sqlite3.c:(.text+0x35714): undefined reference to `log'
/usr/lib/gcc-cross/aarch64-linux-gnu/11/../../../../aarch64-linux-gnu/bin/ld: sqlite3.c:(.text+0x35764): undefined reference to `log'

**Comment by tankeco** [comment]
The problem is how to set `CMAKE_FIND_ROOT_PATH` in `CMakeToolchain` as asked in https://github.com/conan-io/conan/issues/9368

sqlite3 failed to find libm when cross compling:
```cmake
if(ENABLE_FTS5 OR ENABLE_MATH_FUNCTIONS)
    find_library(MATH_LIBRARY m)
    if(MATH_LIBRARY)  # <--- NOT_FOUND
        target_link_libraries(${PROJECT_NAME}  PRIVATE ${MATH_LIBRARY})
    endif()
endif()
```
Add `set(CMAKE_FIND_ROOT_PATH /usr/aarch64-linux-gnu)` before `find_library` fixed the problem.
But I do **not** know how to set `CMAKE_FIND_ROOT_PATH` in `conan_toolchain.cmake`

to reproduce:
```
docker run -it --rm conanio/gcc8-armv8 /bin/bash
pip install conan -U
conan profile new x86 --detect
conan install sqlite3/3.39.4@ --build -pr:b x86 -pr:h default
```

**Comment by memsharded** [comment]
To summarize:

- This is an issue with **cross compiling**, but it doesn't happen while building in native linux (ConanCenter does it for many different Linux configurations)
- This could be a recipe issue, not necessarily a Conan issue, but we need to investigate. For example the ``zlib/1.2.12`` recipe works with the above instructions, so maybe it is just a ``sqlite3`` recipe issue?
- Even if the ``set(CMAKE_FIND_ROOT_PATH /usr/aarch64-linux-gnu)`` could be defined some way in the ``CMakeToolchain``, it is something that doesn't make sense to be hardcoded in the ``conanfile.py`` recipe either. We need to understand better if there is some other way to specify it, I have tried defined the sysroot with:
```bash
$ conan install sqlite3/3.39.4@ --build -pr:b x86 -pr:h default -c tools.build:sysroot=/usr/aarch64-linux-gnu
...
 The C compiler

    "/usr/bin/aarch64-linux-gnu-gcc-8"

  is not able to compile a simple test program.
```

**Comment by prince-chrismc** [comment]
@yushb0602 could you please share which generators you were using and some snippets about hoe you were including sqlite3 in your project?

**Comment by yushb0602** [comment]
> @yushb0602 could you please share which generators you were using and some snippets about hoe you were including sqlite3 in your project?


my conanfile.txt is:
```
[requires]
boost/1.80.0
fmt/9.1.0
spdlog/1.10.0
nlohmann_json/3.11.2
sqlite_orm/1.7.1
openssl/1.0.2s
magic_enum/0.8.1

[generators]
cmake_find_package
cmake
virtualenv
[options]
#boost:without_stacktrace=True
```


I build the libraries with:
`conan install ../cppsrc --profile=./on_aarch64 --build=missing`

After add libm to sqlite manually,  all libraries work for me

**Comment by uilianries** [comment]
@yushb0602 please, **ALWAYS** use build context when cross-building. It's documented here: https://docs.conan.io/en/latest/systems_cross_building/cross_building.html?highlight=cross#conan-v1-24-and-newer

**Comment by memsharded** [comment]
In any case, just to clarify, I am able to reproduce the failure with @tankeco docker instructions, it seems something is missing there, but I am not sure if in the recipe or in Conan.

**Comment by jcar87** [comment]
Thanks @tankeco  for your instructions, and @yushb0602  for reporting the issue.

I believe the issue here has to do with the `CMakeLists.txt` of sqlite3 that lives in Conan Center, rather that Conan itself, and the particulars of the libc math library (`libm.so`/`libm.a`). 

Some notes:
* `undefined reference` linker errors typically mean there is a missing or misplaced `-l` linker statement telling the linker to consider the symbols from a particular library.
* `libm` is part of the libc and `libc` is _typically_ part of the compiler toolchain when cross-building - for historical reasons one needs to add `-lm` when requiring symbols from the maths library, but it not usual to use `find_library` in cmake for this library
* The compiler toolchain in the `conanio/gcc8-armv8` already has the correct libm, and the compiler itself can find it without issues implicitly. The compiler is _already_ using a sysroot, and the sysroot is already correct on a Debian multiarch like this docker image. You can try invoking `$CC -v hello.c -o hello_world` inside  the container, which will show something like this:
```
COLLECT_GCC_OPTIONS='-v' '-o' 'hello_world' '-mlittle-endian' '-mabi=lp64'
 /usr/lib/gcc-cross/aarch64-linux-gnu/8/collect2 -plugin /usr/lib/gcc-cross/aarch64-linux-gnu/8/liblto_plugin.so -plugin-opt=/usr/lib/gcc-cross/aarch64-linux-gnu/8/lto-wrapper -plugin-opt=-fresolution=/tmp/ccVRbUie.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --sysroot=/ --build-id --eh-frame-hdr --hash-style=gnu --as-needed -dynamic-linker /lib/ld-linux-aarch64.so.1 -X -EL -maarch64linux --fix-cortex-a53-843419 -pie -z now -z relro -o hello_world /usr/lib/gcc-cross/aarch64-linux-gnu/8/../../../../aarch64-linux-gnu/lib/../lib/Scrt1.o /usr/lib/gcc-cross/aarch64-linux-gnu/8/../../../../aarch64-linux-gnu/lib/../lib/crti.o /usr/lib/gcc-cross/aarch64-linux-gnu/8/crtbeginS.o -L/usr/lib/gcc-cross/aarch64-linux-gnu/8 -L/usr/lib/gcc-cross/aarch64-linux-gnu/8/../../../../aarch64-linux-gnu/lib/../lib -L/lib/../lib -L/usr/lib/aarch64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc-cross/aarch64-linux-gnu/8/../../../../aarch64-linux-gnu/lib /tmp/ccjksxp1.o -lgcc --push-state --as-needed -lgcc_s --pop-state -lc -lgcc --push-state --as-needed -lgcc_s --pop-state /usr/lib/gcc-cross/aarch64-linux-gnu/8/crtendS.o /usr/lib/gcc-cross/aarch64-linux-gnu/8/../../../../aarch64-linux-gnu/lib/../lib/crtn.o
```

One of the link directories (`-L`) that the compilers passes to the linker is 
`/usr/lib/gcc-cross/aarch64-linux-gnu/8/../../../../aarch64-linux-gnu/lib/../lib` - which happens to contain `libm.so` which is what we need. 

* In _this_ case, in this Docker image, there is no need to pass `CMAKE_SYSROOT` to CMake. This has two main consequences:
 * It will pass `--sysroot` to the compiler - this will cause the compiler to stop looking in the directory it was configured to look into, and look elsewhere for the system and libc libraries. This is perfectly fine for relocatable compiler toolchains or toolchains intended to work with a reference root file system. It is not the case here when we are in a multiarch debian that is already properly configured.
 * It will cause CMake to re-root the entire search path - this might have unintended consequences and possibly interfere with the `find_package` logic that is set up by the `conan_toolchain.cmake` file generated by `CMakeToolchain`. In some cases it may be fine.
 
 


The main issue is how `libm` is searched for and linked in the `CMakeLists.txt` file for `sqlite3`, which lives in Conan center:
* Since `libm` can be considered "part" of the compiler - the way this is typically searched for is *not* `find_library`, but running a small test case to see if the compiler can resolve `-lm` on its own and note that. The `CheckLibraryExists` module can be used for this (docs [here](https://cmake.org/cmake/help/latest/module/CheckLibraryExists.html)).

An example here:
https://stackoverflow.com/questions/34625627/how-to-link-to-the-c-math-library-with-cmake

For older versions of CMake perhaps a different workaround can be used:
https://cmake.org/pipermail/cmake/2019-March/069168.html

If it was a different library that is _not_ already part of the compiler toolchain, `find_library` would be fine. I would note however that issuing a `target_link_libraries` without ensuring that the file was found should probably be corrected as it otherwise lets the error propagate all the way to the linker phase. Additionally, without the proper safeguards, it could also accidentally find the `libm` from the build machine rather than the intended target (or `host` as we call it). 



TL;DR: the `CMakeLists.txt` needs to determine the availability of the math library a different way, here:
https://github.com/conan-io/conan-center-index/blob/master/recipes/sqlite3/all/CMakeLists.txt#L153-L156

If you note a few lines above, the threads library is also searched via `find_package`. It does not use `find_library` internally, but rather does similar checks against the compiler itself: https://github.com/Kitware/CMake/blob/master/Modules/FindThreads.cmake
